### PR TITLE
drivers: flash: stm32h7: Data to write might not be nbytes aligned

### LINKED ISSUE #59029: subsys: lorawan: nvm: STM32H7 unaligned memory access (read) exception when writing LoRaWAN setting names to internal flash
**Describe the bug**
On STM32H7 MCUs, an unaligned memory access exception happens on successful LoRaWAN network join.

When joining a LoRaWAN network by calling `lorawan_join()` function, the LoRaWAN subsystem writes some settings to NVS. Each NVS setting has a string representing its name (a key). NVS setting name strings are treated as `const char` arrays, which go in `.rodata` section. Since there is no alignment requirement for `const char` arrays, linker puts the setting name strings from `lorawan_nvm_settings.c` file one after another (sequentially) in `.rodata` section (there are 7 different LoRaWAN settings in [lorawan_nvm_settings.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/lorawan/nvm/lorawan_nvm_settings.c#L35) file).

In `lorawan_nvm_save_settings` function, LoRaWAN setting name string is passed to `settings_save_one()`, which eventually gets passed to STM32 flash driver's `flash_stm32_write(const struct device *dev, off_t offset, const void *data, size_t len)` function as `const void *data` parameter.
This function then calls STM32H7 flash driver's `flash_stm32_write_range(const struct device *dev, unsigned int offset, const void *data, unsigned int len)` function which casts pointer to data from `const void*` to `const uint64_t*` and passes the data to `write_nwords(const struct device *dev, off_t offset, const uint64_t *data, uint8_t n)` function. 

Finally, after all these calls, pointer which points to arbitrary aligned data (`const char` array located in flash) ends up being casted to `const uint64_t* data` and gets accessed by pointer dereference in `write_ndwords` function:
```
	/* Perform the data write operation at the desired memory address */
	for (i = 0; i < n; ++i) {
		flash[i] = data[i];

		/* Flush the data write */
		__DSB();

		wait_write_queue(&sector);
	}
```
(here is a [link to the line where this code is located](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/flash_stm32h7x.c#L346))

This dereference causes unaligned memory access exception on STM32H7, which could be fixed either in `lorawan_nvm_settings.c` file by copying the setting name string to a temporary 8-byte aligned buffer which will then get passed to `settings_save_one()` function or it can be fixed in `flash_stm32h7x.c` file (better solution) by handling a special case if source data (data to be written to flash) is coming from flash region (e.g. `.rodata` section) and it is unaligned.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using?
   - Custom board with STM32H7A3VI MCU (relevant parts of device tree provided at the end of this report), but any STM32H7A3/B3/B0 MCU is affected, since they use STM32H7 flash driver (`flash_stm32h7x.c`).
   (NVS subsystem cannot be used on STM32H7 MCUs other than STM32H7A3/B3/B0 since their flash blocks are too big for NVS subsystem - 128 KB)
 - What have you tried to diagnose or workaround this issue?
   - A tested temporary workaround for this issue is to copy the setting name string to temporary RAM buffer in `lorawan_nvm_save_settings` function and pass pointer to that buffer when calling `settings_save_one` function instead of passing pointer to strings located in `.rodata` (i.e. located in flash). A proper solution would be to handle unaligned source data located in flash in STM32H7 flash driver.
 - Is this a regression? If yes, have you been able to "git bisect" it to a specific commit?
  - Not a regression, it seems like `write_ndwords()` function in STM32H7 flash driver has always been like this

**To Reproduce**
I encountered this on a custom board, but the issue could be reproduced by building and running `samples/subsys/lorawan/class_a` sample on `stm32h7b3i_dk` board with `storage` flash partition and LoRa transceeiver chip added to board device tree, settings subsystem with NVS backend enabled (`CONFIG_NVS=y CONFIG_SETTINGS=y CONFIG_SETTINGS_NVS=y`) and LoRaWAN NVM enabled (`CONFIG_LORAWAN_NVM_SETTINGS=y`). This should cause unaligned memory access exception to happen upon successful LoRaWAN network join.

**Expected behavior**
STM32H7 flash driver should support unaligned source data to be written to flash, regardless if source data is located in RAM or flash (to be honest, I didn't check if STM32H7 flash driver supports writing unaligned source data from RAM to flash, only from flash (`.rodata`) to flash).

**Impact**
Showstopper for using LoRaWAN subsystem with NVM support (`CONFIG_LORAWAN_NVM_SETTINGS=y`) on STM32H7A3/B3/B0 MCUs. Some LoRaWAN servers don't support joining if you don't send a monotonically increasing counter value on every new join request, so having `CONFIG_LORAWAN_NVM_SETTINGS=y` is a must-have.

**Logs and console output**

Log output - exception happens when stepping over [this line](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/flash_stm32h7x.c#L346) (sorry for bunch of dropped messages, it is because debug log level for flash driver is enabled): 
```
*** Booting Zephyr OS build zephyr-v3.3.0 ***
I: Starting bootloader
I: Primary image: magic=bad, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: none
I: Swap type: none
I: Bootloader chainload address offset: 0x18000
I: Jumping to the first image slot
[00:00:00.000,000] <dbg> flash_stm32h7: stm32h7_flash_init: Flash initialized. BS: 16
[00:00:00.000,000] <dbg> flash_stm32h7: stm32h7_flash_init: Block 0: bs: 8192 count: 256
[00:00:00.000,000] <dbg> flash_stm32h7: flash_stm32h7_write_protection: Disable write protection
*** Booting Zephyr OS build zephyr-v3.3.0 ***
--- 1009 messages dropped ---
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73648, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 65632, len: 2
[00:00:00.621,000] <dbg> lorawan_nvm: lorawan_nvm_data_restore: Restoring LoRaWAN settings
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73552, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73568, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73584, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73600, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73616, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73632, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73648, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73664, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 65600, len: 18
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73552, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73568, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73584, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 65776, len: 1
[00:00:00.621,000] <dbg> lorawan_nvm: on_setting_loaded: Key: Crypto
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73552, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73568, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73584, len: 8
[00:00:00.621,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 65776, len: 56
[00:00:00.621,000] <dbg> lorawan_nvm: lorawan_nvm_data_restore: Crypto version: 16777472, DevNonce: 3, JoinNonce: 1925374
[00:00:00.621,000] <dbg> lorawan_nvm: lorawan_nvm_data_restore: LoRaWAN context restored
[00:00:02.700,000] <dbg> lorawan: lorawan_join: Network join request sent!
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_write_protection: Enable write protection
--- 45 messages dropped ---
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_write: Write offset: 73536, len: 16
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_write_protection: Disable write protection
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_write_protection: Enable write protection
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73520, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73536, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73552, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73568, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73584, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73600, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73616, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73632, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73648, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73664, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73680, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 73696, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_read: Read offset: 98288, len: 8
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_write: Write offset: 65968, len: 16
[00:00:11.022,000] <dbg> flash_stm32h7: flash_stm32h7_write_protection: Disable write protection
[00:00:11.022,000] <err> os: ***** USAGE FAULT *****
[00:00:11.022,000] <err> os:   Unaligned memory access
[00:00:11.022,000] <err> os: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0xffffffff
[00:00:11.022,000] <err> os: r3/a4:  0x00000002 r12/ip:  0x00002b0e r14/lr:  0x0802e92b
[00:00:11.022,000] <err> os:  xpsr:  0x61070000
[00:00:11.022,000] <err> os: Faulting instruction address (r15/pc): 0x0802ebd8
[00:00:11.022,000] <err> os: >>> ZEPHYR FATAL ERROR 31: Unknown error on CPU 0
[00:00:11.022,000] <err> os: Current thread: 0x24002678 (sysworkq)
[00:00:11.094,000] <err> os: Halting system
```

**Environment (please complete the following information):**
 - OS: (e.g. Linux, MacOS, Windows): Debian 12
 - Toolchain (e.g Zephyr SDK, ...): Zephyr SDK 0.16.0
 - Commit SHA or Version used:  tested on v3.3.0 and current main branch (d17d28aa58122c7309d88c9815b5e1462ae596f1)

**Additional context**
`prj.conf` file used with a custom board and project:
```
# Logging
CONFIG_LOG=y
CONFIG_NVS_LOG_LEVEL_WRN=y
CONFIG_LORA_LOG_LEVEL_WRN=y
CONFIG_LORAWAN_LOG_LEVEL_DBG=y
CONFIG_FLASH_LOG_LEVEL_DBG=y

# Drivers
CONFIG_GPIO=y
CONFIG_SERIAL=y
CONFIG_SPI=y
CONFIG_FLASH=y
CONFIG_FLASH_PAGE_LAYOUT=y
CONFIG_FLASH_MAP=y
CONFIG_MPU_ALLOW_FLASH_WRITE=y
CONFIG_LORA=y
CONFIG_LORA_SX12XX=y
CONFIG_REBOOT=y

# Standard C library configuration
CONFIG_NEWLIB_LIBC=y
CONFIG_CBPRINTF_FP_SUPPORT=y

# Kernel configuration
CONFIG_MAIN_STACK_SIZE=4096
CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=4096
CONFIG_ISR_STACK_SIZE=4096

# LoRaWAN subsystem
CONFIG_LORAWAN=y
CONFIG_LORAMAC_REGION_EU868=y
CONFIG_LORAWAN_NVM_SETTINGS=y
#CONFIG_LORAWAN_NVM_NONE=y

# Settings subsystem
CONFIG_NVS=y
CONFIG_SETTINGS=y
CONFIG_SETTINGS_NVS=y
CONFIG_SETTINGS_NVS_SECTOR_COUNT=4
CONFIG_SETTINGS_NVS_SECTOR_SIZE_MULT=1

# Debugging - install VS Code Embedded Tools extension to see Zephyr thread info while debugging
CONFIG_DEBUG_THREAD_INFO=y
CONFIG_INIT_STACKS=y
CONFIG_THREAD_STACK_INFO=y
```

Relevant parts of custom board DTS:
```
/dts-v1/;
#include <st/h7/stm32h7a3Xi.dtsi>
#include <st/h7/stm32h7a3v(g-i)tx-pinctrl.dtsi>
#include <zephyr/dt-bindings/gpio/gpio.h>
#include <zephyr/dt-bindings/clock/stm32h7_clock.h>

/*********************************************************
 *                      ROOT NODE                        *
 *********************************************************/
/ {
    model = "CUSTOM board";
    compatible = "custom, board";

    chosen {
        zephyr,console = &debug_uart;
        zephyr,shell-uart = &debug_uart;
        zephyr,sram = &sram0;
        zephyr,flash = &flash0;
        zephyr,dtcm = &dtcm;
        zephyr,itcm = &itcm;
        zephyr,code-partition = &slot0_partition;
    };

    aliases {
        lora0 = &lora0;
    };
};

/*********************************************************
 *                        CLOCKS                         *
 *********************************************************/
/* HSE clock used for PLL */
&clk_hse {
    status = "okay";
    clock-frequency = <DT_FREQ_M(25)>;
};

/* PLL1P used for system clock (280 MHz) */
&pll {
    status = "okay";
    clocks = <&clk_hse>;
    div-m = <5>;
    mul-n = <112>;
    div-p = <2>;
};

&rcc {
    status = "okay";
    clocks = <&pll>;
    clock-frequency = <DT_FREQ_M(280)>;
    d1cpre = <1>;
    hpre = <1>;
    d1ppre = <2>;
    d2ppre1 = <2>;
    d2ppre2 = <2>;
    d3ppre = <2>;
};

/*********************************************************
 *                  USART CONTROLLERS                    *
 *********************************************************/
debug_uart: &uart7 {
    status = "okay";
    current-speed = <115200>;
    pinctrl-0 = <&uart7_rx_pb3 &uart7_tx_pb4>;
    pinctrl-names = "default";
};

/*********************************************************
 *                    SPI CONTROLLERS                    *
 *********************************************************/
&spi4 {
    status = "okay";
    pinctrl-0 = <&spi4_sck_pe12 &spi4_miso_pe13 &spi4_mosi_pe14>;
    pinctrl-names = "default";
    cs-gpios = <&gpiod 11 GPIO_ACTIVE_LOW>;

    lora0: lora@0 {
        status = "okay";
        compatible = "semtech,sx1272";
        reg = <0>;
        spi-max-frequency = <3000000>;
        reset-gpios = <&gpioe 15 GPIO_ACTIVE_HIGH>;
        dio-gpios = <&gpiob 12 GPIO_ACTIVE_HIGH>,
                    <&gpiod 8 GPIO_ACTIVE_HIGH>,
                    <&gpiod 9 GPIO_ACTIVE_HIGH>,
                    <&gpioe 10 GPIO_ACTIVE_HIGH>,
                    <&gpioe 8 GPIO_ACTIVE_HIGH>;
        power-amplifier-output = "pa-boost";
    };
};

/*********************************************************
 *                  FLASH PARTITIONS                     *
 *********************************************************/
&flash0 {
    partitions {
        compatible = "fixed-partitions";
        #address-cells = <1>;
        #size-cells = <1>;

        /* 64KB for bootloader */
        boot_partition: partition@0 {
            label = "mcuboot";
            reg = <0x00000000 0x00010000>;
            read-only;
        };

        /* storage: 32KB for settings */
        storage_partition: partition@10000 {
            label = "storage";
            reg = <0x00010000 0x00008000>;
        };

        /* application image slot: 912KB */
        slot0_partition: partition@18000 {
            label = "image-0";
            reg = <0x00018000 0x000E4000>;
        };

        /* backup slot: 912KB */
        slot1_partition: partition@FC000 {
            label = "image-1";
            reg = <0x000FC000 0x000E4000>;
        };

        /* swap slot: 128KB */
        scratch_partition: partition@1E0000 {
            label = "image-scratch";
            reg = <0x001E0000 0x00020000>;
        };
    };
};


```

### PR DESCRIPTION
**Pull request by erwango**
Deal with the case of non-nbytes-aligned write input buffer.

Fixes #59029

### CODE REVIEW

File: drivers/flash/flash_stm32h7x.c [hunk 1]
```diff
@@ -368,6 +368,25 @@ int flash_stm32_write_range(const struct device *dev, unsigned int offset,
 	const uint8_t ndwords = FLASH_NB_32BITWORD_IN_FLASHWORD / 2;
 	const uint8_t nbytes = FLASH_NB_32BITWORD_IN_FLASHWORD * 4;
 	uint8_t unaligned_datas[nbytes];
+	uint8_t unaligned_bytes = (uintptr_t)data % nbytes;
+
+	/* First check if *data is nbytes aligned */
+	if (unaligned_bytes != 0) {
+		/* *data is not nbytes byte aligned */
+		memset(unaligned_datas, 0xff, nbytes);
```
Comment (erwango) [inline drivers/flash/flash_stm32h7x.c:376]: Not fully correct:
- I might erase existing written bytes here.
- I'm adding and offset to the expected writing address

**Comment by Laczen** [comment]
@erwango, if the read is unaligned it will never align and it would be needed to memcopy inside the loop. The simplest would be just to always do that.

BTW it is strange that an unaligned end is accepted, shouldn't this be errored out in the range check ?

**Comment by martincalsyn** [comment]
I think this is a broader problem than just the stm32h7 driver. I am encountering the same on stm32u585xx and it's probably not just limited to those. Do we need to change each driver or is there a place we can make a broader change?

**Review by de-nordic — COMMENTED** [review]
I do not think this is correct fix.
NVS, where the problem occurs in reported isseu uses write alignment it reads from flash driver (`write_block_size`).
This means that NVS gets incorrect info on alignment.
Instead of fixing this by overwriting data several times, as I suspect happens here, correct write alignment should be reported for subsystems writing to the flash.

**Comment by erwango** [comment]
> I do not think this is correct fix.

Fully agreed

> NVS, where the problem occurs in reported isseu uses write alignment it reads from flash driver (write_block_size).
This means that NVS gets incorrect info on alignment
> Instead of fixing this by overwriting data several times, as I suspect happens here, correct write alignment should be reported for subsystems writing to the flash.

On this point, I'd be glad you were right, but this is set as:
https://github.com/zephyrproject-rtos/zephyr/blob/0773cc88c881bce51ae510ec150f03b69501e6d1/dts/arm/st/h7/stm32h7a3.dtsi#L18-L20
which is inline with ref man ([RM0455.pdf](https://www.st.com/resource/en/reference_manual/rm0455-stm32h7a37b3-and-stm32h7b0-value-line-advanced-armbased-32bit-mcus-stmicroelectronics.pdf)):
`Flash memory programming by 128 bits (user area) and 16 bits (OTP area)`

Am I missing something?

**Comment by Laczen** [comment]
I think I am now missing what us happening. The problem is not that the flash operation is unaligned, the problem is that the data is unaligned.

The data that is passed to a flash operation should not be aligned, if the flash driver needs this to be aligned it has to do that itsself.

**Comment by Laczen** [comment]
A little test program:
```
#include <zephyr/kernel.h>
#include <zephyr/device.h>
#include <zephyr/drivers/flash.h>
#include <zephyr/devicetree.h>

#define FL_OFFSET 0xff800
#define FL_SIZE 0x200
static const struct device *flash_device =
	DEVICE_DT_GET_OR_NULL(DT_CHOSEN(zephyr_flash_controller));

int main(void)
{
	int rc, i;
	char tst_data[32] = "aaaaaaaaaaaaaaaabbbbbbbbbbbbbbbb";
	printk("Hello World! %s\n", CONFIG_BOARD);

	rc = flash_erase(flash_device, FL_OFFSET, FL_SIZE);
	printk("Flash erase %s\n", rc == 0 ? "[OK]" : "[ERR]");
	for (i = 0; i < 16; i++) {
		rc = flash_write(flash_device, FL_OFFSET + i * 16, &tst_data[i], 16);
		if (rc != 0) {
			break;
		}
	}

	printk("Flash write %s at %i\n", rc == 0 ? "[OK]" : "[ERR]", i);
	return 0;
}
```
This errors for i=1 (on stm32h747i_disco_m7).

**Comment by tmilkovic51** [comment]
@erwango I have tested two of the possible solutions to this issue on STM32H7 series, so please take a look and see if any of these are worth implementing as a fix upstream:

- Use `memcpy` to copy the source data into temporary `write-block-size`-aligned buffer, and then write it to flash: https://github.com/zephyrproject-rtos/zephyr/compare/main...tmilkovic51:zephyr:fix-h7-unaligned-read-in-flash-driver
  - This approach adds an additional space overhead (write block size temporary buffer on stack) and processing overhead (extra `memcpy`-ing)
- Use `UNALIGNED_GET` macro to make compiler take care of unaligned read access to source buffer: https://github.com/zephyrproject-rtos/zephyr/compare/main...tmilkovic51:zephyr:fix-h7-unaligned-read-in-flash-driver-v2
  - This approach adds a slight performance penalty while accessing the source buffer (probably negligible compared to duration of flash write)
  - I'm also not sure if `UNALIGNED_GET` macro is available on non-GNU toolchains (it is defined in `gcc.h`) - other STM32 flash drivers use it (G0, L4 and WB), so compatibility with other toolchains could already be broken for these series


The same issue could also affect G4, L5, U5 and H5 flash drivers (I didn't check it on actual hardware, just inspected the code): 
- [G4](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/flash_stm32g4x.c#L246)
- [L5/U5/H5](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/flash_stm32l5x.c#L345)

**Comment by de-nordic** [comment]
> @erwango I have tested two of the possible solutions to this issue on STM32H7 series, so please take a look and see if any of these are worth implementing as a fix upstream:
> 
>     * Use `memcpy` to copy the source data into temporary `write-block-size`-aligned buffer, and then write it to flash: [main...tmilkovic51:zephyr:fix-h7-unaligned-read-in-flash-driver](https://github.com/zephyrproject-rtos/zephyr/compare/main...tmilkovic51:zephyr:fix-h7-unaligned-read-in-flash-driver)
>       

`memcpy` to intermediate is only needed for unaligned source. In all other cases you can just copy directly from source address. This would give you two paths of execution and increased flash size, but aligned wrights would be faster.

**Comment by Laczen** [comment]
> > @erwango I have tested two of the possible solutions to this issue on STM32H7 series, so please take a look and see if any of these are worth implementing as a fix upstream:
> > ```
> > * Use `memcpy` to copy the source data into temporary `write-block-size`-aligned buffer, and then write it to flash: [main...tmilkovic51:zephyr:fix-h7-unaligned-read-in-flash-driver](https://github.com/zephyrproject-rtos/zephyr/compare/main...tmilkovic51:zephyr:fix-h7-unaligned-read-in-flash-driver)
> > ```
> 
> `memcpy` to intermediate is only needed for unaligned source. In all other cases you can just copy directly from source address. This would give you two paths of execution and increased flash size, but aligned wrights would be faster.

I would skip the trial to optimize and standardize on `UNALIGNED_GET()`, if compilers do not support it it should be added (it is also used in soc_flash_nrf.c). If on some devices it is not needed to do this the macro could be modified to do nothing (for that device).

**Comment by aurel32** [comment]
>     * This approach adds an additional space overhead (write block size temporary buffer on stack) and processing overhead (extra `memcpy`-ing)

There is already a buffer for that called `unaligned_datas` so there is no additional space overhead. I don't think there will be a measurable overhead, given the flash write time is around ~5000 CPU cycles. IMHO it's best to keep the code simple (which also means smaller) and avoid too many code paths.

**Comment by de-nordic** [comment]
What `UNALIGNED_GET` basically does is read data, pointer by pointer, through intermediate variable of type the pointer provided to `UANLIGED_GET` is; so if you just go down to:
https://github.com/zephyrproject-rtos/zephyr/blob/e1efafa31db942f1a2557b967a22da522c4526ee/drivers/flash/flash_stm32h7x.c#L344-L352
And replace
```
flash[i] = data[i]; 
```
with:
```
uint64_t temp = data[i];
flash[i] = temp;
```
that should do the same thing, if `temp` gets aligned on stack to its type.

**Comment by erwango** [comment]
Superseded by #61549

**Pull request closed without merging**